internal/server/boltdbstate/pipeline_run.go

```diff
@@ -0,0 +1,366 @@
+package boltdbstate
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync/atomic"
+
+	"github.com/hashicorp/go-memdb"
+	bolt "go.etcd.io/bbolt"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+	"github.com/hashicorp/waypoint/pkg/server/ptypes"
+)
+
+var pipelineRunBucket = []byte("pipeline_run")
+
+func init() {
+	dbBuckets = append(dbBuckets, pipelineRunBucket)
+	dbIndexers = append(dbIndexers, (*State).pipelineRunIndexInit)
+	schemas = append(schemas, pipelineRunIndexSchema)
+}
+
+// PipelineRunPut creates or updates the given PipelineRun.
+func (s *State) PipelineRunPut(pr *pb.PipelineRun) error {
+	memTxn := s.inmem.Txn(true)
+	defer memTxn.Abort()
+
+	err := s.db.Update(func(dbTxn *bolt.Tx) error {
+		if pr.Pipeline == nil {
+			return status.Error(codes.FailedPrecondition,
+				"A pipeline ref for the pipeline run is required")
+		}
+
+		if pr.Id == "" {
+			id, err := ulid()
+			if err != nil {
+				return err
+			}
+			pr.Id = id
+		}
+
+		// only alter sequence if this is a new pipeline run
+		if pr.Status == pb.PipelineRun_PENDING {
```

Comment:
I wonder if we could simplify this and instead rely on if no `Id` was passed in, it's a new pipeline run and thus a new sequence id? Maybe this is more specific though so I understand this way too